--- sagernet-sing-box/option/group.go	2025-09-17 10:51:57.458756274 +0800
+++ yelnoo-sing-box/option/group.go	2025-09-17 10:51:49.709955970 +0800
@@ -3,16 +3,24 @@
 import "github.com/sagernet/sing/common/json/badoption"
 
 type SelectorOutboundOptions struct {
-	Outbounds                 []string `json:"outbounds"`
-	Default                   string   `json:"default,omitempty"`
-	InterruptExistConnections bool     `json:"interrupt_exist_connections,omitempty"`
+	GroupCommonOption
+	Default                   string `json:"default,omitempty"`
+	InterruptExistConnections bool   `json:"interrupt_exist_connections,omitempty"`
 }
 
 type URLTestOutboundOptions struct {
-	Outbounds                 []string           `json:"outbounds"`
+	GroupCommonOption
 	URL                       string             `json:"url,omitempty"`
 	Interval                  badoption.Duration `json:"interval,omitempty"`
 	Tolerance                 uint16             `json:"tolerance,omitempty"`
 	IdleTimeout               badoption.Duration `json:"idle_timeout,omitempty"`
 	InterruptExistConnections bool               `json:"interrupt_exist_connections,omitempty"`
 }
+
+type GroupCommonOption struct {
+	Outbounds       []string          `json:"outbounds"`
+	Providers       []string          `json:"providers"`
+	Exclude         *badoption.Regexp `json:"exclude,omitempty"`
+	Include         *badoption.Regexp `json:"include,omitempty"`
+	UseAllProviders bool              `json:"use_all_providers,omitempty"`
+}
